# Midterm ML
--------

> Created by: *Josue Arellano Barba*.
>
> Date: *September 14th, 2023*


>Code of Logistical Regression applied in the dataset of London City's Airbnbs.

>Dataset recoved from: http://data.insideairbnb.com/united-kingdom/england/london/2023-06-08/visualisations/listings.csv

In [63]:
## Importing libraries and reading the datatset "listings.csv"

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from imblearn.over_sampling import SMOTE

df = pd.read_csv("listings.csv")
df

id                                               name  \
0                  306333  Rental unit in London · ★4.32 · 2 bedrooms · 3...   
1                  154221  Rental unit in London · 2 bedrooms · 2 beds · ...   
2                  306405  Rental unit in Dagenham · ★4.0 · 1 bedroom · 1...   
3                  157714  Townhouse in Hammersmith · ★4.71 · 1 bedroom ·...   
4                   13913  Rental unit in Islington · ★4.80 · 1 bedroom ·...   
...                   ...                                                ...   
81787  908448542533162051  Rental unit in Greater London · ★New · 1 bedro...   
81788  908448851903630292  Rental unit in Greater London · ★New · 2 bedro...   
81789  908450386316337239  Rental unit in Greater London · ★New · 1 bedro...   
81790  908452889848079702  Rental unit in Greater London · ★New · 3 bedro...   
81791  908453389034894825  Rental unit in Greater London · ★New · 2 bedro...   

         host_id   host_name  neighbourhood_group           neighbourhood  \
0        1573870      Robert                  NaN                 Hackney   
1         211018      Sylvia                  NaN             Westminster   
2        1575800        Mary                  NaN    Barking and Dagenham   
3         757377      Eilinn                  NaN  Hammersmith and Fulham   
4          54730       Alina                  NaN               Islington   
...          ...         ...                  ...                     ...   
81787  314162972  Blueground                  NaN             Westminster   
81788  314162972  Blueground                  NaN                 Lambeth   
81789   36404387       Gizem                  NaN                 Hackney   
81790  518621554       Ahsan                  NaN             Westminster   
81791  504311587      Stefan                  NaN               Greenwich   

        latitude  longitude        room_type  price  minimum_nights  \
0      51.527480  -0.081720  Entire home/apt    156               2   
1      51.489700  -0.139110  Entire home/apt     83               2   
2      51.540720   0.152460     Private room     41               4   
3      51.489740  -0.222080     Private room     69               4   
4      51.568610  -0.112700     Private room     79               1   
...          ...        ...              ...    ...             ...   
81787  51.531046  -0.176821  Entire home/apt    169              90   
81788  51.499511  -0.114164  Entire home/apt    170              30   
81789  51.557216  -0.067643     Private room     40               2   
81790  51.515650  -0.179570  Entire home/apt    392               3   
81791  51.484432   0.002025  Entire home/apt    169               5   

       number_of_reviews last_review  reviews_per_month  \
0                     20  2017-05-31               0.25   
1                      1  2016-01-12               0.01   
2                      3  2012-05-27               0.02   
3                    175  2023-05-02               1.21   
4                     41  2022-12-11               0.26   
...                  ...         ...                ...   
81787                  0         NaN                NaN   
81788                  0         NaN                NaN   
81789                  0         NaN                NaN   
81790                  0         NaN                NaN   
81791                  0         NaN                NaN   

       calculated_host_listings_count  availability_365  \
0                                   1                 4   
1                                   1                 0   
2                                   1               365   
3                                   1                67   
4                                   2               353   
...                               ...               ...   
81787                             280                 1   
81788                             280               272   
81789                               1       

In [64]:
# Dropping the dataset in variables 'X' and 'y'(Targets)
X = df.drop(["neighbourhood"], axis = 1)
y = df["neighbourhood"].values

In [65]:
# Split 'X' and 'y' in train/test.

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [66]:
X_train.shape, X_test.shape, y_test.shape, y_test.shape

((65433, 17), (16359, 17), (16359,), (16359,))

In [67]:
X_train.columns

Index(['id', 'name', 'host_id', 'host_name', 'neighbourhood_group', 'latitude',
       'longitude', 'room_type', 'price', 'minimum_nights',
       'number_of_reviews', 'last_review', 'reviews_per_month',
       'calculated_host_listings_count', 'availability_365',
       'number_of_reviews_ltm', 'license'],
      dtype='object')

In [68]:
# Removing the columns that aren't relevant to the analysis in 'X_train' and 'y_test'.

X_train = X_train.drop(["id", "name", "host_name", "neighbourhood_group", "last_review", "reviews_per_month", "license"], axis = 1)
X_test = X_test.drop(["id", "name", "host_name", "neighbourhood_group", "last_review", "reviews_per_month", "license"], axis = 1)

In [69]:
X_train

host_id   latitude  longitude        room_type  price  \
6465    19400336  51.515260  -0.179900  Entire home/apt     80   
3601    24758981  51.546160  -0.036250  Entire home/apt     75   
62458  448909739  51.570992   0.184980     Private room     52   
6913    40803777  51.462650   0.006090     Private room     55   
38213    1797656  51.496300  -0.133150  Entire home/apt    700   
...          ...        ...        ...              ...    ...   
6265    13954777  51.452220  -0.062150     Private room     35   
54886  315820320  51.542110  -0.148510  Entire home/apt     85   
76820  212707056  51.534383   0.073886  Entire home/apt    125   
860      3050788  51.541470  -0.055760  Entire home/apt    120   
15795   67485192  51.577020  -0.214750     Private room     52   

       minimum_nights  number_of_reviews  calculated_host_listings_count  \
6465                4                 11                               1   
3601                5                  0                               1   
62458               3                  5                               3   
6913                3                 51                               2   
38213               7                  3                              10   
...               ...                ...                             ...   
6265                1                  2                               1   
54886               7                  2                             108   
76820               2                  0                               2   
860                 3                115                               1   
15795               1                 58                               1   

       availability_365  number_of_reviews_ltm  
6465                  0                      0  
3601                  0                      0  
62458               342                      5  
6913                  0                     11  
38213               344                      0  
...                 ...                    ...  
6265                  0                      0  
54886               131                      2  
76820               355                      0  
860                  92                     18  
15795                 0                      3  

[65433 rows x 10 columns]

In [70]:
X_test

host_id   latitude  longitude        room_type  price  \
12600   35530083  51.526850  -0.062690  Entire home/apt     79   
25813    4855233  51.543680  -0.052860  Entire home/apt    103   
36851  222400913  51.562240  -0.008350  Entire home/apt     70   
75771  507730432  51.519966  -0.062601     Private room     45   
38631   20800292  51.533700  -0.101010     Private room     75   
...          ...        ...        ...              ...    ...   
64636    2125620  51.520200  -0.150920  Entire home/apt    595   
65912  482482921  51.626240  -0.054360     Private room     49   
45470   74167394  51.531260  -0.122230     Private room     80   
12418    9066630  51.513760  -0.121590  Entire home/apt    160   
6439    41578570  51.526230  -0.075900     Private room     20   

       minimum_nights  number_of_reviews  calculated_host_listings_count  \
12600               2                 47                               1   
25813               3                 78                               2   
36851              14                  6                              41   
75771               2                  0                               4   
38631               4                  0                               2   
...               ...                ...                             ...   
64636               3                  7                              95   
65912               1                  0                               3   
45470               1                  3                             106   
12418              30                  7                              17   
6439                5                  0                               1   

       availability_365  number_of_reviews_ltm  
12600                 0                      0  
25813                 8                     21  
36851                55                      0  
75771                25                      0  
38631                 0                      0  
...                 ...                    ...  
64636               325                      7  
65912               365                      0  
45470                16                      1  
12418               125                      2  
6439                  0                      0  

[16359 rows x 10 columns]

## Standarization

In [71]:
# Standarizating the column 'room_type' from string to integer

encoder = LabelEncoder()
X_train["room_type"] = encoder.fit_transform(X_train["room_type"])
X_test["room_type"] = encoder.transform(X_test["room_type"])

## Scalling using MinManx

In [72]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

## LogisticRegression Model

In [73]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(max_iter=1000)
model

LogisticRegression(max_iter=1000)

## SMOTE

In [74]:
# Oversample minority class
sm = SMOTE(random_state=42)
X_train_oversampled, y_train_oversampled = sm.fit_resample(X_train, y_train)

# Training

In [75]:
# Train model
model.fit(X_train_oversampled, y_train_oversampled)
score_test = model.score(X_test, y_test)
print("Score en test: ", score_test*100)

Score en test:  73.93483709273183


## Performance metrics

In [76]:
from sklearn.metrics import classification_report

y_pred = model.predict(X_test)

# Calculate and print classification report
classification_rep = classification_report(y_test, y_pred, target_names=list(df["neighbourhood"].unique()))
print("Classification Report in Test:\n", classification_rep)

Classification Report in Test:
                         precision    recall  f1-score   support

               Hackney       0.78      0.83      0.81       109
           Westminster       0.90      0.86      0.88       410
  Barking and Dagenham       0.67      0.95      0.78        84
Hammersmith and Fulham       0.62      0.88      0.73       543
             Islington       0.70      0.97      0.81       151
Kensington and Chelsea       0.77      0.60      0.67      1078
             Southwark       0.10      0.88      0.18        93
               Lambeth       0.91      0.93      0.92       288
                Camden       0.87      0.92      0.89       431
            Wandsworth       0.80      0.98      0.88       140
            Hillingdon       0.75      0.75      0.75       363
         Tower Hamlets       0.80      0.78      0.79      1148
                Barnet       0.64      0.64      0.64       725
  Richmond upon Thames       0.72      0.89      0.79       468
       

| Clase | Precisión | Recall | F1-score | Soporte |
|---|---|---|---|---|
| Hackney | 0.78 | 0.83 | 0.81 | 109 |
| Westminster | 0.90 | 0.86 | 0.88 | 410 |
| Barking and Dagenham | 0.67 | 0.95 | 0.78 | 84 |
| Hammersmith and Fulham | 0.62 | 0.88 | 0.73 | 543 |
| Islington | 0.70 | 0.97 | 0.81 | 151 |
| Kensington and Chelsea | 0.77 | 0.60 | 0.67 | 1078 |
| Southwark | 0.10 | 0.88 | 0.18 | 93 |
| Lambeth | 0.91 | 0.93 | 0.92 | 288 |
| Camden | 0.87 | 0.92 | 0.89 | 431 |
| Wandsworth | 0.80 | 0.98 | 0.88 | 140 |
| Hillingdon | 0.75 | 0.75 | 0.75 | 363 |
| Tower Hamlets | 0.80 | 0.78 | 0.79 | 1148 |
| Barnet | 0.64 | 0.64 | 0.64 | 725 |
| Richmond upon Thames | 0.72 | 0.89 | 0.79 | 468 |
| Greenwich | 0.64 | 0.91 | 0.75 | 105 |
| Haringey | 0.85 | 0.94 | 0.89 | 85 |
| Merton | 0.88 | 0.93 | 0.90 | 171 |
| Brent | 0.83 | 0.76 | 0.79 | 251 |
| Newham | 0.74 | 0.60 | 0.66 | 874 |
| Waltham Forest | 0.68 | 0.71 | 0.69 | 1110 |
| Hounslow | 0.80 | 0.96 | 0.87 | 110 |
| Croydon | 0.72 | 0.81 | 0.76 | 900 |
| Lewisham | 0.70 | 0.83 | 0.76 | 472 |
| Enfield | 0.70 | 0.91 | 0.79 | 288 |
| City of London | 0.76 | 0.81 | 0.78 | 413 |
| Ealing | 0.80 | 0.75 | 0.77 | 156 |
| Bromley | 0.76 | 0.73 | 0.75 | 226 |
| Havering | 0.82 | 0.62 | 0.71 | 918 |
| Harrow | 0.75 | 0.90 | 0.82 | 63 |
| Kingston upon Thames | 0.91 | 0.63 | 0.74 | 1297 |
| Redbridge | 0.80 | 0.92 | 0.85 | 290 |
| Bexley | 0.81 | 0.71 | 0.76 | 793 |
| Sutton | 0.89 | 0.64 | 0.74 | 1806 |


| Resumen |
|---|---|
| Precisión | 0.78 |
| Recall | 0.74 |
| F1-score | 0.75 |
| Accuracy | 0.74 |